geoutils/raster/interpolate.py

```diff
@@ -0,0 +1,329 @@
+from __future__ import annotations
+
+from typing import Any, Callable, Literal, overload
+
+import numpy as np
+import rasterio as rio
+from scipy.interpolate import RectBivariateSpline, RegularGridInterpolator
+from scipy.ndimage import binary_dilation, distance_transform_edt, map_coordinates
+
+from geoutils._typing import NDArrayNum, Number
+from geoutils.raster.georeferencing import _coords, _outside_image, _res, _xy2ij
+
+method_to_order = {"nearest": 0, "linear": 1, "cubic": 3, "quintic": 5, "slinear": 1, "pchip": 3, "splinef2d": 3}
+
+
+def _dist_nodata_spread(order: int, dist_nodata_spread: Literal["half_order_up", "half_order_down"] | int) -> int:
+    """
+    Derive distance of nodata spreading based on interpolation order.
+
+    :param order: Interpolation order.
+    :param dist_nodata_spread: Spreading distance of nodata, either half-order rounded up (default), rounded down, or
+        fixed integer.
+    """
+
+    if dist_nodata_spread == "half_order_up":
+        dist_nodata_spread = int(np.ceil(order / 2))
+    elif dist_nodata_spread == "half_order_down":
+        dist_nodata_spread = int(np.floor(order / 2))
+
+    return dist_nodata_spread
+
+
+def _interpn_interpolator(
+    coords: tuple[NDArrayNum, NDArrayNum],
+    values: NDArrayNum,
+    fill_value: Number = np.nan,
+    bounds_error: bool = False,
+    dist_nodata_spread: Literal["half_order_up", "half_order_down"] | int = "half_order_up",
+    method: Literal["nearest", "linear", "cubic", "quintic", "slinear", "pchip", "splinef2d"] = "linear",
+) -> Callable[[tuple[NDArrayNum, NDArrayNum]], NDArrayNum]:
+    """
+    Create SciPy interpolator with nodata spreading. Default is spreading at distance of half the method order
+    rounded up (i.e., linear spreads 1 nodata in each direction, cubic spreads 2, quintic 3).
+
+    Gives the exact same result as scipy.interpolate.interpn, and allows interpolator to be re-used if required (
+    for speed).
+    In practice, returns either a NaN-modified RegularGridInterpolator or a NaN-modified RectBivariateSpline object,
+    both expecting a tuple of X/Y coordinates to be evaluated.
+
+    Adapted from:
+    https://github.com/scipy/scipy/blob/44e4ebaac992fde33f04638b99629d23973cb9b2/scipy/interpolate/_rgi.py#L743.
+    """
+
+    # REMOVED (native NaN support not reliable)
+    # Adding masking of NaNs for methods not supporting it
+    # method_support_nan = method in ["nearest", "linear"]
+
+    # Derive distance to spread nodata to depending on method order
+    order = method_to_order[method]
+    d = _dist_nodata_spread(order=order, dist_nodata_spread=dist_nodata_spread)
+
+    # If NaNs are not supported
+    if True:
+        # We compute the mask and dilate it to the distance to spread nodatas
+        mask_nan = ~np.isfinite(values)
+        if d != 0:
+            new_mask = binary_dilation(mask_nan, iterations=d).astype("uint8")
```

Comment:
at this stage, it is assumed that the array is 2D no? However, this is not indicated in the docstring and there is no check that this is actually the case I believe. Until now, I assumed `values` could be a 1D array.

Reply:
I leave those checks to SciPy, as this is a non-public subfunction. The public function `Raster.interp_points` always extracts the right shape based on `.coords()` and `get_nanarray`, so there should never be any error there.
To answer the shape: the function works with the same shape for all arrays as described in SciPy. So passing 1D coords tuple and 1D values, or 2D coords tuple and 2D values, and supposedly for ND.
We only test 1D in GeoUtils, and the 2D in xDEM with the new `_reproject_horizontal_xxx` function.

Reply:
Correction: values is always 2D, but the point coordinates can have any 1D/2D/ND shape!